prov/efa/src/efa_mr.c

```diff
@@ -560,6 +563,191 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
+			 existing_mr->ibv_mr->addr,
+			 existing_mr->peer.iface,
+			 existing_mr->ibv_mr->length);
+		return -FI_ENOKEY;
+	}
+
+	err = efa_mr_is_cuda_memory_freed(existing_mr, &cuda_memory_freed);
+	if (err)
+		return err;
+
+	if (!cuda_memory_freed) {
+		/* The same key was assigned to two valid cuda memory region,
+		 * there is no way we can recover from this situation, return error code */
+		EFA_WARN(FI_LOG_DOMAIN, "key %ld has already assigned to another cuda buffer: %p length: %ld\n",
+			 existing_mr->mr_fid.key,
+			 existing_mr->ibv_mr->addr,
+			 existing_mr->ibv_mr->length);
+		return -FI_ENOKEY;
+	}
+
+	EFA_INFO(FI_LOG_DOMAIN, "key %ld has been assigned to cuda buffer: %p length: %ld, which has since been freed\n",
+		 existing_mr->mr_fid.key,
+		 existing_mr->ibv_mr->addr,
+		 existing_mr->ibv_mr->length);
+
+	/* this can only happen when MR cache is enabled, hence the assertion */
+	assert(efa_mr->domain->cache);
+	pthread_mutex_lock(&mm_lock);
+	ofi_mr_cache_notify(efa_mr->domain->cache, existing_mr->ibv_mr->addr, existing_mr->ibv_mr->length);
+	pthread_mutex_unlock(&mm_lock);
+
+	/* due to MR cache's deferred de-registration, ofi_mr_cache_notify() only move the region to dead_region_list
+	 * ofi_mr_cache_flush() will actually remove the region from cache.
+	 * lru is a list of regions that are still active, so we set flush_lru to false.
+	 */
+	ofi_mr_cache_flush(efa_mr->domain->cache, false /*flush_lru */);
+
+	/*
+	 * When MR cache removes a MR, it will call its delete_region() call back. delete_region() calls efa_mr_dereg_impl(),
+	 * which should remove the staled entry from MR map. So insert again here.
+	 */
+	ofi_genlock_lock(&efa_mr->domain->util_domain.lock);
+	err = ofi_mr_map_insert(&efa_mr->domain->util_domain.mr_map, mr_attr,
+				&efa_mr->mr_fid.key, &efa_mr->mr_fid);
+	ofi_genlock_unlock(&efa_mr->domain->util_domain.lock);
+	if (err) {
+		EFA_WARN(FI_LOG_MR,
+			"Unable to add MR to map, even though we already tried to evict staled memory registration."
+			"errno: %d errmsg: (%s) key: %ld buff: %p hmem_iface: %d len: %zu\n",
+			err,
+			fi_strerror(-err),
+			efa_mr->mr_fid.key,
+			mr_attr->mr_iov->iov_base,
+			mr_attr->iface,
+			mr_attr->mr_iov->iov_len);
+		return err;
+	}
+
+	return 0;
+}
+#else
+static
+int efa_mr_update_domain_mr_map(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr)
+{
+	int err;
+
+	mr_attr->requested_key = efa_mr->mr_fid.key;
+	ofi_genlock_lock(&efa_mr->domain->util_domain.lock);
+	err = ofi_mr_map_insert(&efa_mr->domain->util_domain.mr_map, mr_attr,
+				&efa_mr->mr_fid.key, &efa_mr->mr_fid);
+	ofi_genlock_unlock(&efa_mr->domain->util_domain.lock);
+	if (err) {
+		EFA_WARN(FI_LOG_MR,
+			"Unable to add MR to map. errno: %d errmsg: (%s) key: %ld buff: %p hmem_iface: %d len: %zu\n",
+			err,
+			fi_strerror(-err),
+			efa_mr->mr_fid.key,
+			mr_attr->mr_iov->iov_base,
+			mr_attr->iface,
+			mr_attr->mr_iov->iov_len);
+		return err;
+	}
+
+	return 0;
+}
+#endif
```

Comment:
I'd like the endif and else to have #HAVE_CUDA so that it's easier to know what each block belongs to

Reply:
added in new revision.